### PR TITLE
feat(refactor): merge note command

### PR DESCRIPTION
**Pull request by hikchoi**
# feat(refactor): merge note command
This PR:
- adds a new command `Dendron: Merge Note`
- enhances Lookup provider so that it can take a list of `preAcceptValidators` when it is created by the lookup provider factory.
  - this will allow lookup-derivative commands (such as this one) to pass in list of a predicate functions that would be executed when the user hits enter in the quickpick.
  - if any of the `preAcceptValidator` fails, lookup will not be accepted until the user selects a valid item from the quickpick
  - In merge note command, this is used to prevent users from merging the active note to itself.

# Pull Request Checklist

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [x] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [x] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

## Docs
- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [ ] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by hikchoi** [comment]
pending:
- [ ] clean up command logic
- [ ] add proxy metrics
- [ ] add tests
- [ ] confirm engine state is sound after the command
- [ ] add documents

**Comment by hikchoi** [comment]
related to: https://github.com/dendronhq/dendron/issues/906

**Comment by hikchoi** [comment]
new proxy metrics props added to airtable ✅

**Comment by hikchoi** [comment]
Will create async on the new `preAcceptValidator` added to lookup provider when this PR is good to go.

**Comment by hikchoi** [comment]
doc updates at https://github.com/dendronhq/dendron-site/pull/609

**Review by jonathanyeung — APPROVED** [review]
Nice! 

Maybe as a later enhancement - should we run this command automatically when a user does 'Move Note' and selects a destination that already exists? (Maybe first prompting them w/ "Would you like to merge these two notes?")

**Comment by hikchoi** [comment]
> Nice!
> 
> Maybe as a later enhancement - should we run this command automatically when a user does 'Move Note' and selects a destination that already exists? (Maybe first prompting them w/ "Would you like to merge these two notes?")

Good question. I did think about the UX / implementation of this as Move note and Merge note are pretty similar ideas on a surface level.

I think first we need to be careful about intent. When users try to use move note, and select an existing note for the destination, do they really want to do that? what are they expecting the end result to be? Or stepping back once more, does this actually happen frequently, and do we know if that action is intentional (they want / expect it to work) or is it a mistake?

Slightly tangential as I know you aren't suggesting we lump the implementation together with Move note, but I do want to break down the move note command and simplify it so that it does one thing. Currently it's taking care of rename note as a special case of move note, and as a result there are extra moving parts in it that could potentially break and make everything that depend on it not work if that happens.

So I prefer move note's capability to be simple and have it not serve as an actual entry / trigger point for another note-level refactoring capability. Though I do see the benefits of what you're suggesting.

My ideal way of doing this would be disallow and notify things that shouldn't happen in one command, and think about a way to give the users sufficient and clear signifiers that _if_ their intention was different, that some other command can do it for them. So for example, instead of how move note just short-circuits when they pick an existing note in lookup, enhance it so that it uses a `preAcceptValidator` that disallows it in the first place, and show a toast that serve as a signifier that _maybe you want to do this instead?_, but not chain it there directly so that it becomes a trigger point for the Merge Note command.

File: packages/plugin-core/src/commands/MergeNoteCommand.ts [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+          .map((backlink) => backlink.from.id)
+          .filter((ent): ent is string => ent !== undefined)
+      )
+    );
+
+    // for each note that needs to be updated,
+    // find all links that need to be updated from end to front.
+    // then update them.
+    let noteChangeEntries: NoteChangeEntry[] = [];
+    await asyncLoopOneAtATime(noteIDsToUpdate, async (id) => {
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/MergeNoteCommand.ts:330]: does this need to be done sequentially vs in parallel?

File: packages/plugin-core/src/commands/MergeNoteCommand.ts [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+    destNote: NoteProps;
+  }) {
+    const ctx = `${this.key}:updateLinkInNote`;
+    const { id, sourceNote, destNote } = opts;
+    const noteToUpdate = await this.extension.getEngine().getNote(id);
+    if (noteToUpdate !== undefined) {
+      const linksToUpdate = noteToUpdate.links.filter(
+        (link) => link.value === sourceNote.fname
+      );
+      const noteToUpdateDeepCopy = _.cloneDeep(noteToUpdate);
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/MergeNoteCommand.ts:236]: in this case since we are intentionally mutating the note body, do we need to clone?

Comment (kevinslin) [reply packages/plugin-core/src/commands/MergeNoteCommand.ts:236]: also `getNote` shoudl already clone the note

File: packages/plugin-core/src/commands/MergeNoteCommand.ts [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    const deleteSourceChanges = await this.deleteSource({ sourceNote });
+
+    const noteChangeEntries = [
+      ...appendNoteChanges,
+      ...updateBacklinksChanges,
+      ...deleteSourceChanges,
+    ];
+
+    // open the destination note
+    await this.extension.wsUtils.openNote(destNote);
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/MergeNoteCommand.ts:404]: do we need to close the source note?

File: packages/plugin-core/src/commands/MergeNoteCommand.ts [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+    sourceNote: NoteProps;
+    destNote: NoteProps;
+  }) {
+    const { sourceNote, destNote } = opts;
+    // grab body from current active note
+    const appendPayload = sourceNote.body;
+
+    // append to end
+    const destBody = destNote.body;
+    const newBody = `${destBody}\n\n${appendPayload}`;
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/MergeNoteCommand.ts:206]: should we have a line here saying the note was merged? if we go with evernote behavior, it would be

```
{body of previous note}
\n
---
# {title of merged note}
\n
{body of new note)
``

**Review by kevinslin — CHANGES_REQUESTED** [review]
left some comments

File: packages/plugin-core/src/commands/MergeNoteCommand.ts (hunk 4 above)
Comment (hikchoi) [reply packages/plugin-core/src/commands/MergeNoteCommand.ts:206]: that's a good idea.

File: packages/plugin-core/src/commands/MergeNoteCommand.ts (hunk 2 above)
Comment (hikchoi) [reply packages/plugin-core/src/commands/MergeNoteCommand.ts:236]: I was being extra to not mutate it before I explicitly call write but I guess this is now taken care of with the recent store / engine improvements 😄

File: packages/plugin-core/src/commands/MergeNoteCommand.ts (hunk 1 above)
Comment (hikchoi) [reply packages/plugin-core/src/commands/MergeNoteCommand.ts:330]: doesn't need to be sequential. good catch!

File: packages/plugin-core/src/commands/MergeNoteCommand.ts (hunk 3 above)
Comment (hikchoi) [reply packages/plugin-core/src/commands/MergeNoteCommand.ts:404]: I initially thought it'd be nice if the deleted note still stuck around in the editor, but it'd be better to close it to be consistent with the behavior of `Dendron: Delete` now that I think about it. Also better to clean up a note in the editor that isn't in the store to prevent the engine going out of sync potentially.

**Comment by kevinslin** [comment]
looks like we have build errors

**Comment by hikchoi** [comment]
> looks like we have build errors

@kevinslin sorry about that. this week's release had a breaking change with the engine (return signature changed for `writeNote`). Rebased / fixed now

**Review by kevinslin — APPROVED** [review]
